src/compiler.ts

```diff
@@ -3052,6 +3052,13 @@ export class Compiler extends DiagnosticEmitter {
           uniqueMap(flow.contextualTypeArguments)
         );
         if (!type) continue;
+        if (type.kind == TypeKind.VOID) {
+          this.error(
+            DiagnosticCode.Variable_declarations_of_type_void_are_forbidden,
+            typeNode.range
+          );
+          continue;
+        }
```

Comment:
Perhaps better use this error?
```ts
this.error(
  DiagnosticCode.Type_0_is_not_assignable_to_type_1,
  declaration.range, type.toString(), "void"
);
```

cc @dcodeIO

Reply:
Or perhaps just "Type expected"?

Reply:
That would imply
```ts
let v: void;
```
is being assigned

Reply:
I think `let v: void` shouldn't forbidden. Example:
```ts
function test<T>(fn: (s: string) => T): T {
  let res: T;
  res = fn("hello");
  return res;
}

test((): void => console.log);
```

Reply:
How would this be handled? Would the compiler handle `void` type declarations or would `void` have a runtime value that binaryen can optimize away?

Reply:
It should handled and resolved as any other type. The only peculiarity of the void type is that only a void value can be assigned to such a variable, i.e. the same variable (which have void type) or the result of a function that returns the void type (like `console.log` for example)

Reply:
Got it. I'll work on something like that when I find the time.